### ISSUE TITLE
Different loss after iteration by same seed with different reduction way of nll_loss

### ISSUE THREAD

**Issue by jimmy-walker** [pytorch/pytorch#49400]
## 🐛 Bug
Hello, I faced a problem about reduction way of nll_loss, which give me extremely different result by same seed in GCN.

I am using GCN to do binary text classification. First, I use class weight in the nll_loss as belowed. It gives me good result. As I used same seed, everytime I rerun the GCN, I can get the same result. I can check the log each iteration of loss is same.
```python
np.random.seed(args.seed)
torch.manual_seed(args.seed)
if args.cuda:
    torch.cuda.manual_seed(args.seed)

x_class = y.sum(dim=0).float()
z_weight = (1.0/x_class)*x_class.sum(0)/2.0
y = y.max(1)[1].type_as(labels)
return F.nll_loss(preds, y, z_weight)
```
But when I changed the nll_loss to reduction "none", although the loss return is also same as above at the beginning, which value is 0.7351 at the first time, it will show different result after multi iteration at the end.
```python
x_class = y.sum(dim=0).float()
z_weight = (1.0/x_class)*x_class.sum(0)/2.0
y = y.max(1)[1].type_as(labels)
loss_pre = F.nll_loss(preds, y, reduction='none')
final_weight = z_weight[y]
return (loss_pre * final_weight).sum() / final_weight.sum()
```
```
#original loss
Epoch: 0001
 | loss: 0.7351
Epoch: 0002
 | loss: 0.6355
Epoch: 0003
 | loss: 3.2794
...
Epoch: 0010
 | loss: 0.6107
...
Epoch: 0200
 | loss: 0.0207

#loss of reduction none
Epoch: 0001
 | loss: 0.7351
Epoch: 0002
 | loss: 0.6355
Epoch: 0003
 | loss: 3.2797
...
Epoch: 0010
 | loss: 0.6097
...
Epoch: 0200
 | loss: 0.0239
```
Is it a bug? Any help will be appreciated.

## Expected behavior

I expected it will print same loss with reduction 'none' as reduction 'mean' default.

## Environment

```
Collecting environment information...
PyTorch version: 1.3.0
Is debug build: False
CUDA used to build PyTorch: 10.1.243
ROCM used to build PyTorch: N/A
OS: Ubuntu 16.04.6 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
Clang version: Could not collect
CMake version: Could not collect
Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 418.74
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3
HIP runtime version: N/A
MIOpen runtime version: N/A
Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] pytorch-lightning==1.0.4
[pip] torch==1.3.0
[pip] torchvision==0.4.1a0+d94043a
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.168                      0  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.14           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] numpy                     1.17.2           py36haad9e8e_0  
[conda] numpy-base                1.17.2           py36hde5b4d6_0  
[conda] pytorch                   1.3.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-lightning         1.0.4                    pypi_0    pypi
[conda] torchvision               0.4.1                py36_cu101    pytorch
```

cc @ezyang @gchanan @zou3519 @bdhirsh

**Comment by ezyang** [comment]
@kurtamohler as our determinism maestro, you want to take a quick look here?

**Comment by kurtamohler** [comment]
Sure, I'll look into it

**Comment by kurtamohler** [comment]
@jimmy-walker, are you using CUDA or CPU tensors?

Based on the output you posted, it looks like you're using `nll_loss`'s backward function for training. The backward function of `nll_loss`'s CUDA implementation is known and PyTorch 1.7 documents this case to be nondeterministic. With 1.7, you can set `torch.set_deterministic(True)` to detect this particular source nondeterminism and throw an error (https://pytorch.org/docs/stable/generated/torch.set_deterministic.html#torch.set_deterministic).

However, if you're using CPU tensors, or if you're not using the backward function, then we will need to investigate more.

Besides the detection of nondeterministic behavior, there is also the question of whether we can or should update `nll_loss` to act deterministically. `nll_loss`'s CUDA backward function is nondeteterministic because of `atomicAdd` usage, according to this comment: https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu#L59

Since `atomicAdd` is the cause of nondeterminism for many CUDA operations, we could look into ways of avoiding it (if that's not already being done).

**Comment by jimmy-walker** [comment]
@kurtamohler thanks your reply
I am using this code without gpu, ie only cpu.
I have tried some change. I think the problem is related to float precision.
Cause when I use `z_weight = (1.0/x_class)*x_class.sum(0)/2.0`, to create z_weight as `tensor([5.1711, 0.5535])`,  it will show the different loss with different reduction.
When I set z_weight as `tensor([10.0, 1.0])`, the printed result of loss with different reduction is same.
But the problem is my situation is dealing with unbalanced classification as you can see.
The z_weight `tensor([5.1711, 0.5535])` give me good result, If I change to  `tensor([10.0, 1.0])`, it return bad result.
Could you find the way to make reduction of none same with reduciton of mean?

**Comment by kurtamohler** [comment]
Hey @jimmy-walker, sorry for the delay, this issue slipped off my radar.

If I can just summarize my interpretation of the issue: You compared training with `nll_loss(..., reduction='mean')` against `nll_loss(..., reduction='none')` followed by your own custom mean calculation. You expected the two different approaches to give the same exact results, since they are mathematically equivalent. However, after several training epochs, their results started to diverge.

I suspect this is a numerical error issue. If your custom method of applying a mean reduction has a different order of operations than the reduction within the `nll_loss` function, then the two are bound to produce different results for some sets of inputs. To show my point, take a look at the script below. This script calculates `nll_loss` three different ways:

1. `reduction='mean'`
2. `reduction='none'`, followed by your custom mean implementation
3. `reduction='none'`, followed by a slightly different custom mean implementation

Then, it calculates the average difference between the custom implementations and the base `reduction='mean'`. The experiment is run 10 times, each with different random inputs.

```
import torch

print('seed avg_loss_diff_0 avg_loss_diff_1')
for seed in range(10):
    torch.manual_seed(seed)
    preds = torch.rand(100, 100)
    y = torch.rand(100, 100)
    x_class = y.sum(dim=0).float()
    z_weight = (1.0 / x_class) * x_class.sum(0) / 2.0 
    target = y.max(1)[1].to(torch.long)

    loss_mean = torch.nn.functional.nll_loss(preds, target, z_weight)

    loss_no_reduce = torch.nn.functional.nll_loss(preds, target, reduction='none')
    final_weight = z_weight[target]

    loss_custom_mean_0 = (loss_no_reduce * final_weight).sum() / final_weight.sum()
    avg_loss_diff_0 = (loss_mean - loss_custom_mean_0).abs().mean()

    loss_custom_mean_1 = (final_weight * loss_no_reduce).sum() * (final_weight.sum().reciprocal())
    avg_loss_diff_1 = (loss_mean - loss_custom_mean_1).abs().mean()
    
    print(f'{seed} {avg_loss_diff_0} {avg_loss_diff_1}')
```

When I run this in my environment, I get the following:

seed | avg_loss_diff_0 | avg_loss_diff_1
-- | -- | --
0 | 5.96E-08 | 5.96E-08
1 | 5.96E-08 | 5.96E-08
2 | 5.96E-08 | 5.96E-08
3 | 1.79E-07 | 1.19E-07
4 | 2.98E-08 | 2.98E-08
5 | 1.49E-07 | 1.49E-07
6 | 0 | 0
7 | 1.49E-07 | 1.49E-07
8 | 0 | 0
9 | 1.19E-07 | 1.19E-07

Notice that all three different methods can produce slightly different results in some cases (for seed 3 in the table above, all three results were different). Sometimes, we can get lucky, and they all produce the same result, as in seed 6 and 8.

In my opinion, this isn't really an actionable issue--it's just an inevitable side-effect of using floating point numbers. If you want to guarantee that your script produces the same result as `nll_loss(..., reduction='mean')`, your best bet is probably to actually use `nll_loss(..., reduction='mean')`.

Side note: I don't think this qualifies as a determinism issue, since nothing here suggests that subsequent executions of `nll_loss(..., reduction='mean')` on CPU can produce different results given identical inputs.

**Comment by kurtamohler** [comment]
@ezyang and @jimmy-walker, does my comment above make sense, and do you agree that it's a non-issue?

**Comment by ezyang** [comment]
Yup, SGTM.
